m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -540,7 +599,7 @@ private boolean elapsedMaxDelaySinceLastFlush() {
         private void drainQueue() throws TException {
             SizedMetric metrics;
 
-            while ((metrics = metricQueue.poll()) != null) {
+            while ((metrics = queue.poll()) != null) {
```

Comment:
Question: previously I think we would hit this code path solely when we've shutdown the overall reporter `isShutdown() == true`. Now there's a path (the exception case) where we're exiting the processor loop, but not the overall reporter, and then calling `shutdown => drainQueue`. I think there may be scenarios where this function doesn't exit, or at least tries to flush out all of the metrics in the queue before it shuts down. It's perhaps not that dangerous (I think it will probably successfully clear the queue and exit most of the time in practice), but it does strike me as weird.

Wdyt about having the old processor simply exit on error without draining the queue, and leave the queue processing to the next processor that spins up?

Reply:
Good catch, previously that was last resort effort of trying to submit metrics. Most of the time it's futile. With the new approach we should skip it and leave it for the after-coming Processor to handle it.

Had to move forward with it unfortunately to make sure we mitigate the risk ASAP, but please feel free to put up a CR addressing it.